### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by kedarbhat**
- Contains fix for #26: The "fix" for this is to move the array assignments for indices above 0 through run-time index variables rather than through integral constant indices (ie 1). I was unable to reproduce the error in MSVC, but Clang gave me the same error.

- Contains a cleaner fix for #19: to avoid both ODR-usage and messy preprocessor conditional compilation, I extracted local `constexpr` variables/constants for `fvar::order_sum`, which are then passed into `std::min`. This removes the ODR -usage issue entirely and keeps the code consistent across C++ standards.

### CODE REVIEW

**Comment by kedarbhat** [comment]
The "fix" for this is to move the array assignments for indices above 0 through run-time index variables rather than through integral constants. I was unable to reproduce the error in MSVC, but Clang gave me the same error.

File: include/boost/math/differentiation/autodiff.hpp [hunk 1]
```diff
@@ -1291,12 +1288,11 @@ fvar<RealType,Order> sqrt(const fvar<RealType,Order>& cr)
...
     root_type powers = 1;
-    derivatives[1] = numerator / *derivatives;
-    for (size_t i=2 ; i<=order ; ++i)
+    for (size_t i=1 ; i<=order ; ++i)
     {
-      numerator *= -0.5 * ((i<<1)-3);
-      powers *= x;
       derivatives[i] = numerator / (powers * *derivatives);
+      numerator *= root_type(-0.5 * (((i+1)<<1)-3));
+      powers *= x;
```
Comment (shaneasd) [inline include/boost/math/differentiation/autodiff.hpp]: Any comment on the performance implications of this? Seems like before optimizations you're performing these *=s redundantly once each where the previous solution didn't. Similarly you're comparing i to order one additional time with this approach. Does this all just get optimized away at the end of the day?

**Comment by shaneasd** [comment]
That was a quick solution to #26!

File: include/boost/math/differentiation/autodiff.hpp (hunk 1 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: There will be one extra multiplication/assignment each for `numerator`  and for `powers`. A conditional checking to see if we’re in the final iteration (so no such operations are needed) can be added. For arithmetic types, the current code would waste a small number of cycles, but it’s suboptimal to leave the code this way. 

Thanks for catching this!

File: include/boost/math/differentiation/autodiff.hpp [hunk 2]
```diff
@@ -994,11 +994,8 @@ template<typename Func>
 fvar<RealType,Order> fvar<RealType,Order>::apply_coefficients(const size_t order, const Func& f) const
 {
   const fvar<RealType,Order> epsilon = fvar<RealType,Order>(*this).set_root(0);
-#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
-  size_t i = std::min(order, order_sum);
-#else
-  size_t i = order < order_sum ? order : order_sum;
-#endif
+  constexpr auto fvar_order_sum = fvar<RealType,Order>::order_sum;
+  size_t i = (std::min)(order, fvar_order_sum);
```
Comment (pulver) [inline include/boost/math/differentiation/autodiff.hpp]: This is a downgrade to the C++17 version. I'd rather keep the ` #ifndef BOOST_NO_CXX17_IF_CONSTEXPR`.

At some point, support for C++11 will be abandoned just as older C++ versions are not supported. When that time comes, it will be much easier if these compiler checks are in place, rather than manually changing these compromised lines back to the more succinct C++17 version.

Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: I would only comment (for context) that being `constexpr`, `fvar_order_sum` should be applied to `std::min` at compile time as if the argument `fvar::order_sum` were being passed directly; in other words, a no-op in terms of allocating space on the stack as would be the case for a run-time automatic variable, since the `fvar_order_sum` isn’t calculated at run-time (being a constant expression instead), and therefore inline-able for any reasonable compiler.

The ODR issue also affects C++14; it only became deprecated in C++17.

The impetus for the change is that users tend to find preprocessor separations confusing, and I saw this as a reasonable compromise which should have cause no performance degradation while improving readability (being only 1 line as opposed to several the preprocessor solution in place), while also avoiding the need for a namespace scope declaration for `fvar::order_sum`.

If you still would prefer the current code, I’ll remove this change from the PR.

File: include/boost/math/differentiation/autodiff.hpp (hunk 1 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: I updated the PR with a conditional around the two highlighted lines. Thanks again for catching this.

File: include/boost/math/differentiation/autodiff.hpp [hunk 3]
```diff
@@ -1291,12 +1288,13 @@ fvar<RealType,Order> sqrt(const fvar<RealType,Order>& cr)
...
     root_type numerator = 0.5;
     root_type powers = 1;
-    derivatives[1] = numerator / *derivatives;
-    for (size_t i=2 ; i<=order ; ++i)
+    for (size_t i=1 ; i<=order ; ++i)
     {
-      numerator *= -0.5 * ((i<<1)-3);
-      powers *= x;
       derivatives[i] = numerator / (powers * *derivatives);
+      if (i < order) {
```
Comment (pulver) [inline include/boost/math/differentiation/autodiff.hpp]: This is another downgrade for C++17.

I guess I'm not getting my point across about not making these kinds of compromises. I'll try a better explanation later.

File: include/boost/math/differentiation/autodiff_cpp11.hpp [hunk 4]
```diff
@@ -230,15 +231,19 @@ template<typename RealType, size_t Order>
 template<typename RootType>
 fvar<RealType,Order>& fvar<RealType,Order>::negate_cpp11(std::true_type, const RootType&)
 {
-  std::for_each(v.begin(), v.end(), [](RealType& r) { r.negate(); });
+  for (auto& r : v) {
+    r.negate();
+  }
```
Comment (pulver) [inline include/boost/math/differentiation/autodiff_cpp11.hpp]: Why this change?

File: include/boost/math/differentiation/autodiff_cpp11.hpp [hunk 5]
```diff
@@ -91,7 +91,7 @@ template<typename... Orders>
 get_type_at<fvar<RealType,Order>,sizeof...(Orders)> fvar<RealType,Order>::derivative(Orders... orders) const
 {
   static_assert(sizeof...(Orders) <= depth, "Number of parameters to derivative(...) cannot exceed fvar::depth.");
-  return at(orders...) * product(boost::math::factorial<root_type>(orders)...);
+  return at(static_cast<std::size_t>(orders)...) * product(boost::math::factorial<root_type>(static_cast<unsigned>(orders))...);
```
Comment (pulver) [inline include/boost/math/differentiation/autodiff_cpp11.hpp]: Why this change?

File: include/boost/math/differentiation/autodiff.hpp (hunk 3 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: Are you referring to `if (i < order)` in particular? If so, I'm not sure what can be done in C++17 as opposed to earlier standards for this line.

File: include/boost/math/differentiation/autodiff_cpp11.hpp (hunk 5 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff_cpp11.hpp]: I made the types that `fvar::at` and `boost::math::factorial` explicit (both MSVC and Clang complain about `orders` not being `unsigned` for `boost::math::factorial` specifically).

File: include/boost/math/differentiation/autodiff_cpp11.hpp (hunk 4 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff_cpp11.hpp]: I did it for readability, since the lambda expression in this case just wrapping the body of the for-range loop (a single call to a member function with a side effect). If you would prefer the `std::for_each`, I can revert.

File: include/boost/math/differentiation/autodiff.hpp (hunk 2 above)
Comment (shaneasd) [reply include/boost/math/differentiation/autodiff.hpp]: for what it's worth, I agree that this is more readable (at no performance cost) than the ifndef. Seems to me that it's better up until the point that you no longer want to support c++14 at which point it because slightly more verbose. Given the tendency (from my observation at least) for boost to support super old c++ versions I would assume that you'd keep c++14 support for a long time.

If you do decide to go with the compiler switch could you please change it from an ifndef to an ifdef. I found the double negative to be quite confusing.

File: include/boost/math/differentiation/autodiff.hpp (hunk 3 above)
Comment (shaneasd) [reply include/boost/math/differentiation/autodiff.hpp]: I think the point is that here the algorithm has been complicated for all compilers in order to compensate for an issue in non-c++17 compilers. If your primary goal is good code for c++17 and your secondary goal is support for c++11 then this is not acceptable. I wonder though if there is a way to make this work in c++11 without at the very least making the code more complicated for the c++17 reader. i.e. if someone who is compiling in c++17 is reading this code then any additional complication to the code would be worsening that c++17 user's experience.

As for this particular change and independent from pulver's point, are you confident that this condition is actually cheaper than the redundant work it protects against? The most expensive thing here is a couple of floating point multiplications. I was under the impression that a jump would be comparatively expensive. At the very least it's a jump which wasn't present in the original version. Full disclosure: I've made no attempt to profile any of the approaches.

File: include/boost/math/differentiation/autodiff_cpp11.hpp (hunk 4 above)
Comment (shaneasd) [reply include/boost/math/differentiation/autodiff_cpp11.hpp]: Interesting reading on the topic: https://www.fluentcpp.com/2018/03/30/is-stdfor_each-obsolete/

Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff_cpp11.hpp]: I’m a big fan of that site. Since readability is primary to me if no alternative holds any particular performance or efficiency advantage over the others, the loop body is sufficiently simple that achieving any level of abstraction beyond what’s there would be extremely difficult, primarily since we’re calling a member function rather than passing a container element into a function (ie negate(r) for each r in v).

I’m not dogmatic on this choice; of Matthew would prefer std::for_each, I’m fine reverting. The substance of these pull requests are merely suggestions to the main author.

File: include/boost/math/differentiation/autodiff.hpp (hunk 3 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: The alternative to making all non-zero array indices non-literal for non-C++17 is to separate the clauses into separate structs (and using tag dispatching or SFINAE), which is perfectly reasonable. I chose this way to deal with the issue you found since it changed the original code the least (by moving the write to index 1 into the loop and updating the two variables within the loop after assignment).

As for the expense of the jump, I was considering a worst than arithmetic root type (ie boost::multiprecision::cpp_bin_float_50), for which the extra multiplication/assignment operations would be less trivial. GCC offers functionality to help with these jumps, but I’m not aware of MSVC/Clang supporting them.

Comment (pulver) [reply include/boost/math/differentiation/autodiff.hpp]: I would much prefer this change:
```
#ifndef BOOST_NO_CXX17_IF_CONSTEXPR
    derivatives[1] = numerator / *derivatives;
#else // for compilers that compile this branch when order=0.
    derivatives[std::min(size_t(1),order)] = numerator / *derivatives;
#endif
```
I don't wish to ever sacrifice (1) performance or (2) code expression/brevity offered by C++17 for the sake of backward compatibility.

Using `#ifndef BOOST_NO_CXX17_IF_CONSTEXPR` may seem ugly, and perhaps seemingly contradictory to (2), but it is a way to preserve the ideal C++17 version so that when the time comes to drop C++11 (and eventually C++14) support, these #conditionals wil make it much easier. It's also a good way to demonstrate the improvement that C++17 has over previous versions.

In this particular case, there is a slight performance degradation in checking the `i < order` conditional, and extra increment `i+1`, in every iteration which is not in the original version.

Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: np, I'll update the PR

File: include/boost/math/differentiation/autodiff.hpp (hunk 2 above)
Comment (pulver) [reply include/boost/math/differentiation/autodiff.hpp]: > If you still would prefer the current code, I’ll remove this change from the PR.

Yes please. My most previous [comment](https://github.com/pulver/autodiff/pull/29/files#r270627779) also applies here. Except that in this case there is indeed no performance degradation, but instead it is the loss of brevity made available to us by C++17.

File: include/boost/math/differentiation/autodiff.hpp [hunk 6]
```diff
@@ -1786,9 +1784,9 @@ struct promote_args_2<differentiation::detail::fvar<RealType0, Order0>,
 {
   using type = differentiation::detail::fvar<typename promote_args_2<RealType0, RealType1>::type,
 #ifndef BOOST_NO_CXX14_CONSTEXPR
-           std::max(Order0,Order1)>;
+    (std::max)(Order0,Order1)>;
```
Comment (pulver) [inline include/boost/math/differentiation/autodiff.hpp]: Curious why the parentheses? `(std::max)` Does it fix a warning?

File: include/boost/math/differentiation/autodiff_cpp11.hpp (hunk 5 above)
Comment (pulver) [reply include/boost/math/differentiation/autodiff_cpp11.hpp]: Thanks. That makes sense for `factorial`. Is the `static_cast<std::size_t>(orders)` necessary to satisfy a warning as well?

File: include/boost/math/differentiation/autodiff_cpp11.hpp (hunk 4 above)
Comment (pulver) [reply include/boost/math/differentiation/autodiff_cpp11.hpp]: That's fine for the `_cpp11.hpp` file - please don't make this change to `autodiff.hpp`.

For what it's worth, the reason I originally started using `std::for_each` was that `cppcheck` complained about this in `operator/=(const root_type& ca)`:

    for (RealType& x : v)
        x /= ca;

Changing it to `std::for_each(v.begin(), v.end(), [&ca](RealType& x) { x /= ca; });` satisfied it. I don't recall off-hand what the complaint was.

File: include/boost/math/differentiation/autodiff.hpp (hunk 6 above)
Comment (kedarbhat) [reply include/boost/math/differentiation/autodiff.hpp]: MSVC requires it, otherwise you can pick up their std::min/std::max macros. Boost's developer guidelines touch on this (more specifically to `std::numeric_limits<>::max()`/`std::numeric_limits<>::min()`). The alternative is to define `NOMINMAX`, so `Windows.h` won't define any of these macros.
[Boost Developer Guidelines](https://github.com/boostorg/boost/wiki/Guidelines:-Warnings)

File: include/boost/math/differentiation/autodiff.hpp (hunk 2 above)
Comment (pulver) [reply include/boost/math/differentiation/autodiff.hpp]: > If you do decide to go with the compiler switch could you please change it from an ifndef to an ifdef. I found the double negative to be quite confusing.

In the short term as a new user of the code I would agree with your point. In the longer term, after getting familiar with this logic, I would prefer the current version.

Conceptually (IMO) it is easier to see the higher-level code which comes closer to expressing the intent of the code block first. When reading through the code I can then mentally skip the `#else` block after understanding the meaning.

File: include/boost/math/differentiation/autodiff.hpp (hunk 6 above)
Comment (pulver) [reply include/boost/math/differentiation/autodiff.hpp]: If it's not required or satisfies a warning, I'd prefer to not bend the code to satisfy Windows.

If there is a warning, an alternative would be to try changing `BOOST_NO_CXX14_CONSTEXPR` to `BOOST_NO_CXX17_IF_CONSTEXPR` in case that helps (just a guess).

File: include/boost/math/differentiation/autodiff.hpp (hunk 3 above)
Comment (shaneasd) [reply include/boost/math/differentiation/autodiff.hpp]: On the subject of the apparent contradiction to (2) have you considered simply ifdefing the entire method and putting the c++11 version in a separate header? That way if you're working in c++17 you can simply ignore that header entirely and all you have to complicate things is a conditional that's always true. I notice you already have a c++11 header. I'm not sure if this is its purpose. Generally speaking it would be bad practice to duplicate an entire method just to tweak a small portion of it but in this case it might be the best option in terms of providing the optimal c++17 experience.

File: include/boost/math/differentiation/autodiff.hpp (hunk 6 above)
Comment (shaneasd) [reply include/boost/math/differentiation/autodiff.hpp]: Windows.h contains this macro ```#define max(a,b)            (((a) > (b)) ? (a) : (b))```
As such if a user were to have this code
```
#include <windows.h>
#include <autodiff.hpp>
```
then the expression ```std::max(Order0,Order1)``` resolves to ```std::(((Order0) > (Order1)) ? (Order0) : (Order1))```
in a sample program I whipped up, this gives the error ```Error	C2589	'(': illegal token on right side of '::'```

This is discussed here: https://stackoverflow.com/questions/11544073/how-do-i-deal-with-the-max-macro-in-windows-h-colliding-with-max-in-std . Kedarbhat's approach is the second answer.

The parentheses around max basically say "this isn't a macro". The macro definition in windows.h is truly awful but not something easily changed. Without the parentheses the options are
- Don't completely support windows
- Require your windows users to be careful about the relative ordering of these two headers
- #undef min/max in your header. This has the unfortunate consequence of affecting everything after the autodiff include
- Force windows users to use the NOMINMAX before they include windows. This may break their code and may be very difficult for them if they're not including windows.h directly (e.g. in the worst case if they're getting it via some third party header which depends on the macro).

Comment (pulver) [reply include/boost/math/differentiation/autodiff.hpp]: Thank you for the clear explanation of the the issue. It seems that from the Boost Developer Guidelines link Kedar provided, it recommends adding the `/DNOMINMAX` compiler switch which I very much agree with.

As you probably saw, I'd like to see what the boost guys say about this problem as well.

Comment (shaneasd) [reply include/boost/math/differentiation/autodiff.hpp]: That document is quite peculiarly structured but that's not the take away message I got from that section at all. Note this paragraph in particular:
> So it is far better to check that clashes are correctly avoided by bracketing as above because others may use your code when MIN/MAX macros are defined rather than using this compiler option.)

Nonetheless this document way well be dated so will be interesting to see what the boots guys say.
